Source/WebCore/workers/service/background-fetch/BackgroundFetch.h

```diff
@@ -0,0 +1,154 @@
+        ~Record();
+
+        void complete(const CreateLoaderCallback&);
+        void abort();
+
+        void setAsCompleted() { m_isCompleted = true; }
+        bool isCompleted() const { return m_isCompleted; }
+
+        uint64_t responseDataSize() const { return m_responseDataSize; }
+        bool isMatching(const ResourceRequest&, const CacheQueryOptions&) const;
+        BackgroundFetchRecordInformation information() const;
+
+        void retrieveResponse(BackgroundFetchStore&, RetrieveRecordResponseCallback&&);
+        void retrieveRecordResponseBody(BackgroundFetchStore&, RetrieveRecordResponseBodyCallback&&);
+
+    private:
+        Record(BackgroundFetch&, BackgroundFetchRequest&&, size_t);
+
+        void didSendData(uint64_t) final;
+        void didReceiveResponse(ResourceResponse&&) final;
+        void didReceiveResponseBodyChunk(const SharedBuffer&) final;
+        void didFinish(const ResourceError&) final;
+
+        WeakPtr<BackgroundFetch> m_fetch;
+        BackgroundFetchRecordIdentifier m_identifier;
+        String m_fetchIdentifier;
+        ServiceWorkerRegistrationKey m_registrationKey;
+        BackgroundFetchRequest m_request;
+        size_t m_index { 0 };
+        ResourceResponse m_response;
+        std::unique_ptr<BackgroundFetchRecordLoader> m_loader;
+        uint64_t m_responseDataSize { 0 };
+        bool m_isCompleted { false };
+        bool m_isAborted { false };
+        Vector<RetrieveRecordResponseCallback> m_responseCallbacks;
+        Vector<RetrieveRecordResponseBodyCallback> m_responseBodyCallbacks;
+    };
+
+    using MatchBackgroundFetchCallback = CompletionHandler<void(Vector<Ref<Record>>&&)>;
+    void match(const RetrieveRecordsOptions&, MatchBackgroundFetchCallback&&);
+
+    bool abort();
+
+    void perform(const CreateLoaderCallback&);
+
+    bool isActive() const { return m_isActive; }
+    ClientOrigin origin() const { return m_origin; }
+    uint64_t downloadTotal() const { return  m_options.downloadTotal; }
+    uint64_t uploadTotal() const { return m_uploadTotal; }
+
+    void unsetRecordsAvailableFlag();
+
+private:
+    void didSendData(uint64_t);
+    void storeResponse(size_t, ResourceResponse&&);
+    void storeResponseBodyChunk(size_t, const SharedBuffer&);
+    void didFinishRecord(size_t, const ResourceError&);
+
+    void recordIsCompleted(size_t);
+    void handleStoreResult(BackgroundFetchStore::StoreResult);
+    void updateBackgroundFetchStatus(BackgroundFetchResult, BackgroundFetchFailureReason);
+
+    void store();
+
+    String m_identifier;
+    Vector<Ref<Record>> m_records;
+    BackgroundFetchOptions m_options;
+    uint64_t m_uploadTotal { 0 };
+    BackgroundFetchResult m_result { BackgroundFetchResult::EmptyString };
+    BackgroundFetchFailureReason m_failureReason { BackgroundFetchFailureReason::EmptyString };
+    bool m_recordsAvailableFlag { true };
```

Comment:
Would be nice to keep the enums and booleans together for better packing.

Reply:
ok